room/integration-tests/testapp/src/androidTest/java/androidx/room/integration/testapp/test/PrepackageTest.java

```diff
@@ -566,6 +571,156 @@ public void openDataDirDatabase_versionZero_badSchema() throws IOException {
+        TestPrepackagedCallback callback = new TestPrepackagedCallback();
+        ProductsDatabase database = Room.databaseBuilder(
+                context, ProductsDatabase.class, "products.db")
+                .createFromAsset("databases/products_v1.db", callback)
+                .build();
+
+        assertThat(callback.mOpenPrepackagedDatabase, is(false));
+
+        database.getProductDao().countProducts();
+
+        assertThat(callback.mOpenPrepackagedDatabase, is(true));
+        database.close();
+    }
+
+    @Test
+    public void onCreateFromFile_calledOnOpenPrepackagedDatabase() throws IOException {
+        Context context = ApplicationProvider.getApplicationContext();
+        context.deleteDatabase("products_external.db");
+        File dataDbFile = new File(ContextCompat.getDataDir(context), "products_external.db");
+        context.deleteDatabase(dataDbFile.getAbsolutePath());
+
+        InputStream toCopyInput = context.getAssets().open("databases/products_v1.db");
+        copyAsset(toCopyInput, dataDbFile);
+
+        TestPrepackagedCallback callback = new TestPrepackagedCallback();
+        ProductsDatabase database = Room.databaseBuilder(
+                context, ProductsDatabase.class, "products_external.db")
+                .createFromFile(dataDbFile, callback)
+                .build();
+
+        assertThat(callback.mOpenPrepackagedDatabase, is(false));
+
+        database.getProductDao().countProducts();
+
+        assertThat(callback.mOpenPrepackagedDatabase, is(true));
+
+        database.close();
+    }
+
+    @Test
+    public void onCreateFromZippedAsset_calledOnOpenPrepackagedDatabase() {
+        Context context = ApplicationProvider.getApplicationContext();
+        context.deleteDatabase("products.db");
+
+        final Callable<InputStream> inputStreamCallable = () -> {
+            ZipInputStream zipInputStream =
+                    new ZipInputStream(context.getAssets().open("databases/products_v1.db.zip"));
+            zipInputStream.getNextEntry();
+            return zipInputStream;
+        };
+
+        TestPrepackagedCallback callback = new TestPrepackagedCallback();
+        ProductsDatabase database = Room.databaseBuilder(
+                context, ProductsDatabase.class, "products.db")
+                .createFromInputStream(inputStreamCallable, callback)
+                .build();
+
+        assertThat(callback.mOpenPrepackagedDatabase, is(false));
+
+        database.getProductDao().countProducts();
+
+        assertThat(callback.mOpenPrepackagedDatabase, is(true));
+
+        database.close();
+    }
+
+    @Test
+    public void onOpenedDbTwice_calledOnPrepackagedCallbackOnce() {
+        Context context = ApplicationProvider.getApplicationContext();
+        context.deleteDatabase("products.db");
+
+        ProductsDatabase db1 = null;
+        ProductsDatabase db2 = null;
+        try {
+            TestPrepackagedCallback callback1 = new TestPrepackagedCallback();
```

Comment:
If we update `TestPrepackagedCallback` to use an int for counting everytime the callback is called then we don't need `callback2` making this slightly simpler.